Add `confirm-short-comments` feature

darkred opened the pull request:
<!--

Thanks for contributing! 🍄 Do not ignore this template, plz.

1. Does this PR close/fix an existing issue? Write something like `Closes #10`

Help us test and visualize this PR:

2. What pages does this PR affect? Include some REAL URLs where you tested the code
3. If applicable, add demonstrative screenshots or gifs

Lastly:

4. Open the PR as draft and review it yourself first. Fix what you find and explain weird code, if necessary.

-->

Closes #3911 

## Test URLs

Any new/existing issue page: e.g.
https://github.com/sindresorhus/refined-github/issues/new
https://github.com/sindresorhus/refined-github/issues/3911


## Screenshot

![111637477-23fc9500-8802-11eb-8c1b-4023b7b1c371](https://user-images.githubusercontent.com/723651/111687809-64c0d200-8833-11eb-9ce2-8b8e0f57eda9.jpg)

fregante commented on source/features/confirm-too-short-comments.tsx:
```diff
@@ -0,0 +1,47 @@
+// function handleComment() {
+function handleComment(event: delegate.Event<MouseEvent, KeyboardEvent, HTMLAnchorElement>): void {
+    let body = document.querySelector('textarea[name="comment[body]"]').value.length;
+    if (body < 3 && !confirm('Body too short, sure?')) {
```
The text wasn’t to be taken literally 😃

darkred replied:
Yes, of course, but it seemed short and straightforward, so I thought to keep it in this initial stage.
I'll rephrase it.

darkred replied:
I changed it into `Less than 3 chrs, are you sure?`.

fregante replied:
That’s not grammatically correct and has a typo

darkred replied:
Ok, I've changed the questions to `Body/Title is less than 3 characters, are you sure?`
I also made separate questions for body and title.

fregante commented on source/features/confirm-too-short-comments.tsx:
```diff
@@ -0,0 +1,47 @@
+
+
+// function handleComment() {
+function handleComment(event: delegate.Event<MouseEvent, KeyboardEvent, HTMLAnchorElement>): void {
```
Can you try merging these 2 functions? They do the same thing. You can merge the selectors in a single call for the textareas and then another check + question for the title, if present

darkred replied:
I merged the selectors together. 
I also combined the checks and question, all into one. Isn't this the same?

fregante replied:
If 2 versions are logically the same, one should prefer the shorter one 😉

darkred replied:
Thank you. 
I made the checks and questions separate for comment and for new issue.

fregante commented:
Ensure tests pass and that it’s mostly done before converting to non-draft

darkred commented:
I fixed all errors expept this one:
`Generic type 'Event' requires between 0 and 2 type arguments.`
for [this line ](https://github.com/sindresorhus/refined-github/pull/4123/commits/c8b16163528471ae6d5dda0db1f17240783600d6#diff-a76df45996e6ce7cc1c7d28ea7ec78e81e7fdca5cac889471cef89ab241dc55cR9)
If I removed all 3 arguments, or 1, it caused other errors.
Unfortunately my Typescript knowledge is limited.

fregante commented on source/features/confirm-too-short-comments.tsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+function init(): void {
+
+	const formComment = 'form.js-new-comment-form';
+	const formNewIssue = 'form.new_issue';
+	
+	if (pageDetect.isIssue()){
+		document.querySelector(formComment)?.addEventListener('submit', handleIssueComment );
+	} else if (pageDetect.isNewIssue()){
+		document.querySelector(formNewIssue)?.addEventListener('submit', handleIssueComment );
+	}
+
```
This should be a single event listener too, added via `delegate`, likely

```js
delegate(document, 'the 2 form selectors', submit, handleIssueComment)

darkred replied:
I hope I got it right: [`a801f49` (#4123)](https://github.com/sindresorhus/refined-github/pull/4123/commits/a801f4954913c7ec9f5d903f35f52c8b010e7c5e#diff-a76df45996e6ce7cc1c7d28ea7ec78e81e7fdca5cac889471cef89ab241dc55cL18-R18)

fregante commented on source/features/confirm-too-short-comments.tsx:
```diff
@@ -0,0 +1,37 @@
+void features.add(__filebasename, {
+    awaitDomReady: false,
+	include: [
+		pageDetect.isNewIssue, // Find which one you need on https://fregante.github.io/github-url-detection/
```
This comment needs to be resolved and removed

fregante commented on source/features/confirm-too-short-comments.tsx:
```diff
@@ -0,0 +1,37 @@
+	let title = document.querySelector('input#issue_title')?.value.length;
+	if ( ((body && body < 3) || (title && title < 3)) && !confirm('Less than 3 chrs, are you sure?') ) {
+			event.preventDefault();
+			event.stopImmediatePropagation(); // I don't know if it's necessary to stop comments, they're sent via ajax
```
This comment is for you to find out, not to be part of the code

darkred replied:
Ok, removed the comment.

fregante replied:
Is the code necessary? Did you try without?

darkred replied:
Yes, it is: I see it's necessary for blocking comments submission.

yakov116 commented on source/features/confirm-short-comments.tsx:
```diff
@@ -5,12 +5,17 @@ import features from '.';
 function handleIssueComment(event: delegate.Event<MouseEvent, KeyboardEvent, HTMLAnchorElement>): void {
-    let body = document.querySelector('textarea[name="comment[body]"], textarea#issue_body')?.value.length;
+    let textareas = document.querySelector('textarea[name="comment[body]"], textarea#issue_body')?.value.length;
 	let title = document.querySelector('input#issue_title')?.value.length;
```
```suggestion
	const textareas = select('textarea[name="comment[body]"], textarea#issue_body')!.value.length;
	const title = select('input#issue_title').value.length;
```

darkred replied:
Thank you.

darkred replied:
It says that "Outdated suggestions cannot be applied" so I committed them manually _(I used optional chaining `?` in the second line, as the `title` textarea might not exist in the page)._
Thank you.

yakov116 commented on source/features/confirm-short-comments.tsx:
```diff
@@ -0,0 +1,32 @@
+	delegate(document, 'form.js-new-comment-form, form.new_issue', 'submit', handleIssueComment)
+}
+
+void features.add(__filebasename, {
+    awaitDomReady: false,
+	include: [
+		pageDetect.isNewIssue,
+		pageDetect.isIssue
+	],
+	init
+});
```
```suggestion
void features.add(__filebasename, {
	include: [
		pageDetect.isNewIssue,
		pageDetect.isIssue
	],
	awaitDomReady: false,
	init
});
```

darkred commented:
I'm having this problem now:
if I press Esc to cancel the issue/comment submission,
then the form becomes disabled (I can no longer type in it),
and I have to reload the page (maybe multiple times) or open it in new tab, to be able to type again.

Is my use of `preventDefault()/stopImmediatePropagation()` wrong?

darkred commented:
@fregante 
I'm having this problem now:
if I press Esc to cancel the issue/comment submission,
then the form itself becomes disabled (I can no longer type in it),
and I have to reload the page (maybe multiple times) or open it in new tab, to be able to type again.

To solve the above, it seems like the code needs two things:
1. `else` branches to both `if`, because of the `confirm()`
2. a way to re-enable the `Submit` buttons that become disabled after Esc .

<details>
<summary>Regarding 1:</summary>

I tried changing using an `else` like e.g.

```
if ( (textareas && textareas < 3) && !confirm('Body is less than 3 characters, are you sure?') ) {
event.preventDefault();
event.stopImmediatePropagation();
} else {
return true;
}
```
but unfortunately the button remains disabled: see [screen capture](https://user-images.githubusercontent.com/723651/111777110-86659c00-88bb-11eb-9808-60375b712ba6.gif)

I also tried 
```
} else {
select('button.btn.btn-primary.btn-block')?.click();
}
```
but I think that's not an optimal solution.

</details>

Regarding 2, I can't find any way to fix that.

darkred commented:
With my latest commit, there's now only `confirm()` containing a message for either or both title/body.

&nbsp;
Unfortunately, there are these issues: 
- In the new issue page:
  - in the title input, I can only submit via <kbd>Ctrl+Enter</kbd> the first time - after that, I can only submit either via <kbd>Enter</kbd>.
  BUT, in body textarea, <kbd>Ctrl+Enter</kbd> still works after first blocking.
  - After blocking a submission, the `Submit` button remains disabled .  
  *The only way I found to manually re-enable the `Submit` button, is to manually delete the contents of the textarea, 
click somewhere away, and re-click inside the element.*

- In an existing issue:
  - after blocking a submission, both the `Close with comment` and `Comment` buttons remain disabled . 
If I now manually delete the contents of the textarea, click somewhere away and re-click inside the textarea and type a letter, then only the `Comment` button becomes enabled,
and I can only submit by clicking `Submit` (the submit shortcut <kbd>Ctrl+Enter</kbd> doesn't work).

The only "fix" I found for the buttons becoming disabled is:
```
const element = document.activeElement;
element?.blur();
element?.focus();
```
but it doensn't seem to help.

fregante commented:
Try dispatching a fake InputEvent or using textFieldEdit.insert()

The CI isn’t running on this branch. Please send a PR from a branch other than main and make sure to run `npm run lint` or `npm run fix` since the tests are failing.

darkred commented:
Here is the new PR: #4141
I used `textFieldEdit.insert()`

Thanks a lot for your guidance and patience 🙂

fregante commented:
> Once you got it working, send a new PR

I see that it still doesn't work as it should. If you can't finish the PR autonomously you're asking us to figure it out with you, which means that the PR isn't useful to us, sorry.

darkred commented:
I understand, and I am the one that should I ask sorry: I shouldn't make the PR in its current status, in the first place. 
It's non-working and it looks more like a StackOverflow question. Sorry about that.

I wish I could make it into an actually helpful contribution.
